src/api/c++/z3++.h

```diff
@@ -394,10 +425,19 @@ namespace z3 {
     class array {
         T *      m_array;
```

Comment:
we can maybe change this to a unique_ptr and let C++ do its magic.

Reply:
I agree but I think we can do better.

Is there a reason we can't just replace this class with `std::vector`?  It seems to do the same thing (well, the resize function looses its daya when resizing so, but maybe that is a bug?). The only thing is that this has a custom constructor for `ast_vector_tpl`, but that is either something we could switch to a custom constructor for vector (sort of like how std encourages people to make custom `std::hash` functors), or we could just make it into a function. I currently do not see the benefit of using `array` over `std::vector`.

Reply:
agreed. I checked the users and they are all super simple. Actually many of them can be replaced with std::vector. Can we keep this class alone (ie, revert changes), and then do another PR for array pls?

Reply:
Ok. After this PR I will make a separate PR replacing classes that could be replaced by stl classes like `std::vector`

Reply:
I take back what I said. This is an internal class and using std::vector is not better. Thus adding move constructors & etc is not needed.
If anything, I would switched this to a unique_ptr.

Reply:
Interesting. Can you expand on why `std;:vector` is not preferred? I seem to be lacking the knowledge as to why it being internal means we want `class array` as opposed to just `std::vector` or `template <typename T> using array = std::vector<T>;` if the desire is to maintain the name 'array'.

Somewhat related, but if we want to keep array, is the resize function buggy? It does resize, but it does not keep the data it had during the resize.

Reply:
If nothing else, why not `class array{ std::vector<T> arr; /* ... */ };` or something like that? What is the benefit of maintaining this class as is (or with the unique pointer).

Reply:
std::vector wastes space vs a simple array. Plus given that this is an internal API, there's no need for move constructors and etc.
If anything, change this to unique_ptr, otherwise leave it as-is.

Reply:
I suppose `class array` indeed does not contain a size. It doesn't seem to be used as a member variable, so I'm not sure that justifies this over a vector. But that would be another PR if it is worth it. I will make this one a `unique_ptr` for now.